providers/common/kdfs/kbkdf.c

```diff
@@ -0,0 +1,303 @@
+typedef struct {
+    void *provctx;
+    PROV_DIGEST digest;
+
+    unsigned char *KI;
+    size_t KI_len;
+    unsigned char *Label;
+    size_t Label_len;
+    unsigned char *Context;
+    size_t Context_len;
+} KBKDF;
+
+/* Definitions needed for typechecking. */
+static OSSL_OP_kdf_newctx_fn kbkdf_new;
+static OSSL_OP_kdf_freectx_fn kbkdf_free;
+static OSSL_OP_kdf_reset_fn kbkdf_reset;
+static OSSL_OP_kdf_derive_fn kbkdf_derive;
+static OSSL_OP_kdf_settable_ctx_params_fn kbkdf_settable_ctx_params;
+static OSSL_OP_kdf_set_ctx_params_fn kbkdf_set_ctx_params;
+
+static void *kbkdf_new(void *provctx)
+{
+    KBKDF *ctx;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    ctx->provctx = provctx;
+    return ctx;
+}
+
+
+static void kbkdf_free(void *vctx)
+{
+    KBKDF *ctx = vctx;
+
+    kbkdf_reset(ctx);
+    OPENSSL_free(ctx);
+}
+
+static void kbkdf_reset(void *vctx)
+{
+    KBKDF *ctx = vctx;
+
+    ossl_prov_digest_reset(&ctx->digest);
+    OPENSSL_clear_free(ctx->Context, ctx->Context_len);
+    OPENSSL_clear_free(ctx->Label, ctx->Label_len);
+    OPENSSL_clear_free(ctx->KI, ctx->KI_len);
+    memset(ctx, 0, sizeof(*ctx));
+}
+
+/* SP800-108 section 5.1. */
+static int kbkdf_derive_counter(EVP_MAC_CTX *ctx_init,
+                                unsigned char *Label, size_t Label_len,
+                                unsigned char *Context, size_t Context_len,
+                                unsigned char *K_i, size_t h, uint32_t L,
+                                unsigned char *KO, size_t KO_len)
+{
+    int ret = 0;
+    EVP_MAC_CTX *ctx = NULL;
+    size_t written = 0, to_write;
+    const unsigned char zero = 0;
+    uint32_t counter, i;
+
+    for (counter = 1; written < KO_len; counter++) {
+        i = htobe32(counter);
+
+        ctx = EVP_MAC_CTX_dup(ctx_init);
+        if (ctx == NULL)
+            goto done;
+
+        if (!EVP_MAC_update(ctx, (unsigned char *)&i, 4) ||
+            !EVP_MAC_update(ctx, Label, Label_len) ||
+            !EVP_MAC_update(ctx, &zero, 1) ||
+            !EVP_MAC_update(ctx, Context, Context_len) ||
+            !EVP_MAC_update(ctx, (unsigned char *)&L, 4))
+            goto done;
+
+        if (!EVP_MAC_final(ctx, K_i, NULL, h))
+            goto done;
+
+        to_write = KO_len - written;
+        memcpy(KO + written, K_i, MIN(to_write, h));
+        written += h;
+
+        EVP_MAC_CTX_free(ctx);
+        ctx = NULL;
+    }
+
+    ret = 1;
+done:
+    EVP_MAC_CTX_free(ctx);
+    return ret;
+}
+
+static int kbkdf_derive(void *vctx, unsigned char *key, size_t keylen)
+{
+    KBKDF *ctx = vctx;
+    OPENSSL_CTX *provctx = PROV_LIBRARY_CONTEXT_OF(ctx->provctx);
+    int ret = 0;
+    const EVP_MD *md;
+    EVP_MAC *mac;
+    const char *mdname;
+    EVP_MAC_CTX *ctx_init;
+    OSSL_PARAM params[3];
+    unsigned char *K_i = NULL;
+    uint32_t L = htobe32(keylen * 8);
```

Comment:
Is this going to work across all versions (There are no other references to this currently in the codebase - it seems to come from endian.h on my local machine).

Reply:
Agreed, our usual endian handling _union_ is more appropriate.
Not that I particularly like it but it is invasive throughout the source code.

Personally, I'd not worry about the _inline_.   A static function is as good these days.